c_emulator/rvfi_dii.cpp

```diff
@@ -124,26 +360,59 @@ void rvfi_handler::get_and_send_packet(packet_reader_fn reader,
   KILL(lbits)(&packet);
 }
 
+void rvfi_handler::send_packet_raw(const void *data, size_t len,
+                                   bool config_print)
+{
+  if (config_print) {
+    fprintf(stderr, "Sending packet with length %zd... ", len);
+  }
+  const size_t send_size = len;
+  if (send_size > 4096) {
+    fprintf(stderr, "Unexpected large packet size (> 4KB): %zd\n", send_size);
+    exit(1);
+  }
+  if (write(dii_sock, data, len) != len) {
+    fprintf(stderr, "Writing RVFI DII trace failed: %s\n", strerror(errno));
+    exit(1);
+  }
+  return;
+}
+
 void rvfi_handler::send_trace(bool config_print)
 {
   if (config_print) {
     fprintf(stderr, "Sending v%d trace response...\n", trace_version);
   }
   if (trace_version == 1) {
-    get_and_send_packet(zrvfi_get_exec_packet_v1, config_print);
+    auto pkg = rvfi_get_exec_packet_v1();
+    send_packet_raw(&pkg, sizeof(decltype(pkg)), config_print);
   } else if (trace_version == 2) {
-    get_and_send_packet(zrvfi_get_exec_packet_v2, config_print);
-    if (zrvfi_int_data_present)
-      get_and_send_packet(zrvfi_get_int_data, config_print);
-    if (zrvfi_mem_data_present)
-      get_and_send_packet(zrvfi_get_mem_data, config_print);
+    auto pkg = rvfi_get_exec_packet_v2();
+    send_packet_raw(&pkg, sizeof(decltype(pkg)), config_print);
+    if (rvfi_int_data_present) {
+      send_packet_raw(&rvfi_int_data, sizeof(decltype(rvfi_int_data)),
+                      config_print);
+    }
+    if (rvfi_mem_data_present) {
+      send_packet_raw(&rvfi_mem_data, sizeof(decltype(rvfi_mem_data)),
+                      config_print);
+    }
   } else {
     fprintf(stderr, "Sending v%d packets not implemented yet!\n",
             trace_version);
     abort();
   }
 }
 
+RVFI_DII_Instruction_Packet RVFI_DII_Instruction_Packet_from_u64(uint64_t value)
```

Comment:
```suggestion
RVFI_DII_Instruction_Packet RVFI_DII_Instruction_Packet_from_u64(uint64_t value)
```
```suggestion
RVFI_DII_Instruction_Packet RVFI_DII_Instruction_Packet::from_u64(uint64_t value)
```

Might be nicer as a static function defined in the struct?